File: C++/Stack/Stack_02.cpp
/*
===============================================================================
Program Name    : Stack - Push Operation
Program Number  : 02
File Name       : Stack_02.cpp
Description     : This program demonstrates the Push() operation of a Stack
                  using C++. The Push operation inserts a new element at the
                  top of the Stack following the LIFO principle.

Date            : August 2026
Language        : C++
Author          : Ishwari Vijaykumar Surve
===============================================================================

Concepts Used:
1. Class
2. Object-Oriented Programming (OOP)
3. Self-Referential Structure
4. Dynamic Memory Allocation
5. Constructor

Key Learning Points:
- Insert an element at the top of the Stack.
- Understand dynamic memory allocation using new.
- Increase the Stack size after insertion.
- Follow the LIFO (Last-In-First-Out) principle.

===============================================================================
*/

#include<iostream>
using namespace std;

#pragma pack(1)

struct node
{
    int data;
    struct node *next;
};

class Stack
{
private:
    struct node *first;
    int iCount;

public:
    Stack();
    void Push(int iNo);
    int Pop();
    int Peep();
    void Display();
    int Count();
};

// Constructor
Stack::Stack()
{
    this->first = NULL;
    this->iCount = 0;
}

// Push (Insert First)
void Stack::Push(int iNo)
{
    struct node *newn = NULL;

    newn = new struct node;

    newn->data = iNo;
    newn->next = NULL;

    if(first == NULL)
    {
        first = newn;
    }
    else
    {
        newn->next = first;
        first = newn;
    }

    iCount++;
}

// Pop
int Stack::Pop()
{
    return 0;
}

// Peep
int Stack::Peep()
{
    return 0;
}

// Display
void Stack::Display()
{

}

// Count
int Stack::Count()
{
    return iCount;
}

// Main
int main()
{
    Stack sobj;

    sobj.Push(11);
    sobj.Push(21);
    sobj.Push(51);
    sobj.Push(101);

    cout<<"Number of nodes are : "<<sobj.Count()<<"\n";
/*
===============================================================================
Expected Output:
===============================================================================

Number of nodes are : 4

===============================================================================
Memory Representation:
===============================================================================

                TOP
                 │
                 ▼
             +-------+
             | 101 | •────┐
             +-------+    │
                          ▼
             +-------+
             |  51 | •────┐
             +-------+    │
                          ▼
             +-------+
             |  21 | •────┐
             +-------+    │
                          ▼
             +-------+
             |  11 | •───► NULL
             +-------+

iCount = 4

===============================================================================
Algorithm:
===============================================================================

1. Create a new node.
2. Store the given data into the node.
3. If the Stack is empty:
      Make first point to the new node.
4. Otherwise:
      Insert the new node before the first node.
      Update first.
5. Increment the Stack count.
6. Stop.

===============================================================================
Program Flow:
===============================================================================

            Start
               │
               ▼
        Create New Node
               │
               ▼
      Is Stack Empty?
        │            │
      Yes           No
       │             │
       ▼             ▼
 first = newn   newn->next = first
                    │
                    ▼
               first = newn
                    │
                    ▼
             Increment Count
                    │
                    ▼
                   Stop

===============================================================================
Time Complexity:
===============================================================================

Push() : O(1)

Count() : O(1)

Space Complexity : O(1)

===============================================================================
Conclusion:
===============================================================================

Program 002 demonstrates the Push() operation of a Stack using C++. The
program inserts new elements at the top of the Stack using dynamic memory
allocation and follows the Last-In-First-Out (LIFO) principle.

===============================================================================
*/
    return 0;
}

File: Java/Stack/Stack_02.java
/*
===============================================================================
Program Name    : Stack - Push Operation
Program Number  : 02
File Name       : Stack_02.java

Description     : This program implements the Push operation of a Stack using
                  a singly linear linked list. The Push operation inserts a
                  new element at the first position of the Stack.

Date            : August 2026
Language        : Java
Author          : Ishwari Vijaykumar Surve
===============================================================================

Concepts Used:
1. Class
2. Object-Oriented Programming (OOP)
3. Self-Referential Class
4. Constructor
5. Singly Linear Linked List
6. Stack
7. Push Operation
8. Dynamic Memory Allocation
9. LIFO Principle

Key Learning Points:
- Understand the Push operation of a Stack.
- Learn how a new node is inserted at the first position.
- Understand how the first pointer is updated.
- Learn how iCount is updated after insertion.
- Understand the LIFO (Last In First Out) principle.

===============================================================================
*/

class node
{
    public int data;
    public node next;
}

class Stack
{
    private node first;
    private int iCount;

    public Stack()
    {
        first = null;
        iCount = 0;
    }

    ///////////////////////////////////////////////////////////////////////////
    //
    // Function Name : Push
    // Description   : Insert element at First
    //
    ///////////////////////////////////////////////////////////////////////////

    public void Push(int iNo)
    {
        node newn = null;

        newn = new node();

        newn.data = iNo;
        newn.next = null;

        newn.next = first;
        first = newn;

        iCount++;
    }
}

class Stack_02
{
    public static void main(String A[])
    {
        Stack sobj = new Stack();

        sobj.Push(11);
        sobj.Push(21);
        sobj.Push(51);
        sobj.Push(101);

        System.out.println("Elements inserted successfully");
    }
}

/*
===============================================================================
Expected Output:
===============================================================================

Elements inserted successfully

===============================================================================
Memory Representation:
===============================================================================

After Push(11):

first
  |
  v
+------+-------+
|  11  | NULL  |
+------+-------+


After Push(21):

first
  |
  v
+------+-------+    +------+-------+
|  21  |   o-------->|  11  | NULL|
+------+-------+    +------+-------+


After Push(51):

first
  |
  v
+------+-------+    +------+-------+    +------+-------+
|  51  |   o-------->|  21  |   o-------->|  11  | NULL|
+------+-------+    +------+-------+    +------+-------+


After Push(101):

first
  |
  v
+------+-------+    +------+-------+    +------+-------+    +------+-------+
| 101  |   o-------->|  51  |   o-------->|  21  |   o-------->|  11  | NULL|
+------+-------+    +------+-------+    +------+-------+    +------+-------+

iCount = 4

===============================================================================
Algorithm:
===============================================================================

1. Start the program.
2. Create a new node dynamically.
3. Store the given element in the data member of the new node.
4. Set the next pointer of the new node to NULL.
5. Connect the new node to the existing first node.
6. Update first to point to the new node.
7. Increment iCount by 1.
8. Repeat the Push operation for all given elements.
9. Display the success message.
10. Stop the program.

===============================================================================
Program Flow:
===============================================================================

        Start
          |
          v
    Create New Node
          |
          v
   Store Given Element
          |
          v
    newn.next = first
          |
          v
      first = newn
          |
          v
      iCount++
          |
          v
    Push Next Element
          |
          v
         Stop

===============================================================================
Time Complexity:
===============================================================================

Push() : O(1)

===============================================================================
Space Complexity:
===============================================================================

O(1) auxiliary space per new node.

O(n) total space for n Stack nodes.

===============================================================================
Conclusion:
===============================================================================

Program 02 demonstrates the Push operation of a Stack using a singly linear
linked list in Java. Each new element is inserted at the first position of
the Stack, and the first pointer is updated accordingly. The program
demonstrates the basic implementation of the LIFO (Last In First Out)
principle.

===============================================================================
*/
